modules/python/python-logmsg.c

```diff
@@ -270,6 +274,83 @@ py_log_message_set_pri(PyLogMessage *self, PyObject *args, PyObject *kwrds)
   Py_RETURN_NONE;
 }
 
+static PyObject *
+_datetime_timestamp(PyObject *py_datetime)
+{
+  PyObject *py_tzinfo = _py_get_attr_or_null(py_datetime, "tzinfo");
+  if (!py_tzinfo)
+    {
+      PyErr_Format(PyExc_ValueError, "Error obtaining tzinfo");
+      return NULL;
+    }
+  PyObject *py_epoch = PyDateTimeAPI->DateTime_FromDateAndTime(1970, 1, 1, 0, 0, 0, 0, py_tzinfo,
+                       PyDateTimeAPI->DateTimeType);
+
+  PyObject *py_delta = _py_invoke_method_by_name(py_datetime, "__sub__", py_epoch,
+                                                 "PyDateTime", "py_datetime_to_logstamp");
+  if (!py_delta)
+    {
+      Py_XDECREF(py_tzinfo);
+      Py_XDECREF(py_epoch);
+      PyErr_Format(PyExc_ValueError, "Error calculating POSIX timestamp");
+      return NULL;
+    }
+
+  PyObject *py_posix_timestamp = _py_invoke_method_by_name(py_delta, "total_seconds", NULL,
+                                                           "PyDateTime", "py_datetime_to_logstamp");
+  Py_XDECREF(py_tzinfo);
+  Py_XDECREF(py_delta);
+  Py_XDECREF(py_epoch);
+
+  return py_posix_timestamp;
+}
+
+static gint
+_datetime_get_gmtoff(PyObject *py_datetime)
+{
+  PyObject *py_utcoffset = _py_invoke_method_by_name(py_datetime, "utcoffset", NULL, "PyDateTime",
+                                                     "py_datetime_to_logstamp");
+  if (!py_utcoffset)
+    {
+      PyErr_Format(PyExc_ValueError, "Error obtaining timezone info");
+      return -1;
```

Comment:
Returning -1 is treated as valid return value at the caller. It would worth signalling the error to the caller.

Reply:
I was under the impression that `PyErr_Format` already does signal for python, based on trial actually it does not make any difference. (which makes me think the other way around). Maybe you have some extra info about it.

Reply:
To throw exception from C for the user, the c callback function needs to return `NULL`. In that case the python machinery does the right thing, for example allowing user to catch the exception.

It is a little tedious work, one needs to make sure to propagate error through the whole call chain, just to finally return NULL.

Reply:
Thanks, kinda strange that it does not allow the user to catch, but raises the issue anyway. Fixed.